### PR TITLE
🚀 4.0 production ready grpc implementation for inter service communication

### LINKED ISSUE #49: 🚀 Epic 4.0: Production-Ready gRPC Implementation for Inter-Service Communication
📝 Description
  Migrate the entire internal communication backbone of the Currency Converter Ecosystem from REST/HTTP to a fully production-ready
  gRPC architecture. This epic establishes a high-performance, strictly contracted, and resilient inter-service communication layer
  capable of handling massive concurrency and enterprise-scale loads.

  ✅ Acceptance Criteria (Epic 4.0)
   - [ ] 🏗️ Centralized Contract Management: All .proto files (auth.proto, main.proto, rate.proto) are versioned, strictly managed,
     and built centrally via common-config.
   - [ ] 🛡️ Zero-Trust Security: Mutual TLS (mTLS) enforced for all internal gRPC traffic.
   - [ ] 🕸️ Service Mesh/Load Balancing: Client-side load balancing and proper integration with Kubernetes/Envoy for gRPC traffic.
   - [ ] 🛡️ Resilience Patterns: Circuit breakers, automatic retries with exponential backoff, and timeouts implemented on all gRPC
     channels (via Resilience4j).
   - [ ] 📊 Observability: Distributed tracing (OpenTelemetry), Prometheus metrics, and centralized logging via gRPC interceptors.
   - [ ] 🚨 Graceful Degradation: Fallback mechanisms established for critical gRPC failures.

  ---

 🛠️ Technical Requirements
   - 📡 Protocol: Protobuf 3
   - 🚀 Server/Client: grpc-spring-boot-starter (Netty-based)
   - 🛡️ Resilience: resilience4j-spring-boot3
   - 📊 Tracing: micrometer-tracing-bridge-otel

  📦 Dependencies
   - 🟢 Spring Boot 3.5+
   - 🔗 io.grpc:grpc-netty-shaded
   - 📦 net.devh:grpc-server-spring-boot-starter

  ⚠️ Impact Assessment
   - 💥 Breaking Changes: Internal infrastructure (K8s ingress/services) must be updated to route HTTP/2 (gRPC) traffic.
   - 📈 Performance: Expect a significant drop in CPU/Memory footprint for serialization, and lower network latency.

  🗓️ Timeline
   - ⏱️ Estimated Effort: 3-4 Weeks
   - 🎯 Priority: Highest (Release 4.0 Core Feature)
   - 📅 Deadline: Before Release 4.0 Feature Freeze

  👨‍💻 Assignee: @adiozdaniel 

  🏷️ Labels
   - grpc
   - production-ready
   - epic
   - performance
   - zero-trust
   - release-4.0

### PR DESCRIPTION
**Pull request by adiozdaniel**
🛠️ Key Architectural Changes

  🏗️ 1. Centralized Contract Management (common-config)
   - Established common-config as the "Single Source of Truth" for the entire ecosystem.
   - Protobuf Definitions: Created versioned .proto files (auth.proto, main.proto, rate.proto) for all service domains.
   - Automated Generation: Configured Maven-driven code generation for Java stubs using Protobuf 4.29.3 and gRPC 1.70.0.

  🔐 2. Auth Service: Identity & Token Validation
   - gRPC Server: Implemented AuthGrpcService exposing Authenticate and ValidateToken RPCs.
   - Reactive Bridge: Integrated gRPC StreamObserver with existing WebFlux Mono logic for non-blocking performance.
   - Quality Fixes: Resolved critical SonarQube type-safety findings related to unchecked list casts in claims processing.

  📡 3. Rate Service: High-Performance Exchange Rates
   - gRPC Server: Implemented RateGrpcService for binary exchange rate fetching.
   - Coverage Hardening: Achieved a strict 85% instruction coverage threshold by enhancing the test suite with comprehensive
     edge-case scenarios.
   - Stability: Resolved test-time port conflicts by implementing dedicated test profiles for gRPC server management.

  🔗 4. Main Service: Edge Transcoder & Zero-Trust Client
   - REST-to-gRPC Transcoding: Transformed main-service into an edge gateway that translates external REST requests into optimized
     internal gRPC calls.
   - Continuous Auth validation: Implemented GrpcAuthFilter, an asynchronous interceptor that validates every request's JWT against
     the auth-service via gRPC before routing.
   - Resilience: Configured Resilience4j Circuit Breakers and Retries on all outgoing gRPC channels.

  ---

  ⚙️ Infrastructure & Engineering Standards
   - Centralized BOM: All gRPC, Protobuf, and Resilience4j versions are now managed in the parent pom.xml to prevent version drift.
   - Connection Tuning: Implemented HTTP/2 multiplexing, keep-alive pings (20s), and multiplexed channels to ensure connection
     stability under high load.
   - Documentation: Fully aligned the project's documentation (README.md files for all modules) to reflect the new gRPC-centric
     architecture.

  ---

  📊 Verification & Quality Results
   - Reactor Build: Successfully verified clean compilation of the entire reactor: common-config -> auth-service -> rate-service ->
     main-service.
   - Test Suite: 
       - Total Tests Passed: 253 tests across all modules.
       - Coverage: All modules meet or exceed the 85% coverage mandate.
   - Security: Verified asynchronous token validation flow between main-service and auth-service.

  💥 Impact
   - Performance: Massive reduction in serialization/deserialization overhead and network latency.
   - Scalability: Capable of handling significantly higher concurrency due to multiplexed gRPC channels.
   - Reliability: Stronger contract enforcement through Protobuf and hardened error-handling mapping.

### CODE REVIEW

File: auth-service/Dockerfile [hunk 1]
```diff
@@ -14,7 +14,7 @@ WORKDIR /app
 ARG SPRING_PROFILES_ACTIVE=development
 ENV SPRING_PROFILES_ACTIVE=${SPRING_PROFILES_ACTIVE}
 
-COPY target/auth-service-0.0.1-SNAPSHOT.jar app.jar
+COPY target/*.jar app.jar
```
[the remark this answers is not in the document]
Comment (adiozdaniel) [reply auth-service/Dockerfile:17]: Thank you for the review, but I think you checked a wrong commit. The current commit reflects a rectified file.

File: rate-service/Dockerfile [hunk 2]
```diff
@@ -20,7 +20,7 @@ ENV SPRING_PROFILES_ACTIVE=${SPRING_PROFILES_ACTIVE}
 
 # Copy the built JAR from target directory
 # NOTE: JAR must be built before running docker build
-COPY target/rate-service-0.0.1-SNAPSHOT.jar app.jar
+COPY target/*.jar app.jar
```
[the remark this answers is not in the document]
Comment (adiozdaniel) [reply rate-service/Dockerfile:23]: Thank you for the review, but I think you checked a wrong commit. The current commit reflects a rectified file.

File: main-service/Dockerfile [hunk 3]
```diff
@@ -13,7 +13,7 @@ WORKDIR /app
 ARG SPRING_PROFILES_ACTIVE=development
 ENV SPRING_PROFILES_ACTIVE=${SPRING_PROFILES_ACTIVE}
 
-COPY target/main-service-0.0.1-SNAPSHOT.jar app.jar
+COPY target/*.jar app.jar
```
[the remark this answers is not in the document]
Comment (adiozdaniel) [reply main-service/Dockerfile:16]: Thank you for the review, but I think you checked a wrong commit. The current commit reflects a rectified file.

File: config-server/Dockerfile [hunk 4]
```diff
@@ -14,7 +14,7 @@ WORKDIR /app
 ARG SPRING_PROFILES_ACTIVE=development
 ENV SPRING_PROFILES_ACTIVE=${SPRING_PROFILES_ACTIVE}
 
-COPY target/config-server-0.0.1-SNAPSHOT.jar app.jar
+COPY target/*.jar app.jar
```
[the remark this answers is not in the document]
Comment (adiozdaniel) [reply config-server/Dockerfile:17]: Thank you for the review, but I think you checked a wrong commit. The current commit reflects a rectified file.

File: config-server/src/main/java/com/currencyconverter/configserver/config/EnvironmentRepositoryConfig.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                configRepoPath + "{application}",
+                configRepoPath
+            });
+            
+            return nativeFactory.build(nativeProperties);
+        } else {
+            // For non-local profiles, return null to let Spring use the default Git repository
+            // The Git repository configuration is handled by application.properties
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (adiozdaniel) [reply config-server/src/main/java/com/currencyconverter/configserver/config/EnvironmentRepositoryConfig.java:48]: Good catch! Though, this is a hack to redirect the config to local repo in development mode

File: config-server/src/main/resources/application.properties [hunk 6]
```diff
@@ -4,7 +4,7 @@
 server.port=8888
 spring.application.name=config-server
 # Default to staging (GitHub)
-spring.profiles.active=local
+spring.profiles.active=staging
```
[the remark this answers is not in the document]
Comment (adiozdaniel) [reply config-server/src/main/resources/application.properties:7]: Thanks for the observation, but this was agreed upon to set the default to staging

File: main-service/src/main/java/com/currencyconverter/mainservice/grpc/RateGrpcClient.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                    .setFrom(from)
+                    .setTo(to)
+                    .build();
+            
+            // Set 5 seconds deadline
+            RateResponse response = rateStub.withDeadline(Deadline.after(5, TimeUnit.SECONDS))
+                    .getRate(request);
+            
+            return BigDecimal.valueOf(response.getRate());
+        });
```
[the remark this answers is not in the document]
Comment (adiozdaniel) [reply main-service/src/main/java/com/currencyconverter/mainservice/grpc/RateGrpcClient.java:52]: Wow! Great catch indeed!
Refactored RateGrpcClient to use FutureStub bridged to Mono via Futures.addCallback, ensuring zero blocking on the WebFlux event loop.

File: main-service/src/main/java/com/currencyconverter/mainservice/grpc/AuthGrpcClient.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    public Mono<ValidateResponse> validateToken(String token) {
+        return Mono.fromCallable(() -> {
+            ValidateRequest request = ValidateRequest.newBuilder()
+                    .setToken(token)
+                    .build();
+            
+            // Set 2 seconds deadline for auth
+            return authStub.withDeadline(Deadline.after(2, TimeUnit.SECONDS))
+                    .validateToken(request);
+        });
```
[the remark this answers is not in the document]
Comment (adiozdaniel) [reply main-service/src/main/java/com/currencyconverter/mainservice/grpc/AuthGrpcClient.java:47]: Refactored AuthGrpcClient to use FutureStub bridged to Mono via Futures.addCallback, ensuring zero blocking on the WebFlux event loop.

File: main-service/README.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                       └──────────────┘
+```
+
+## 📡 API Endpoints
+
+### Currency Conversion
+
+`POST /api/v1/convert`
+
+Converts an amount from one currency to another.
```
[the remark this answers is not in the document]
Comment (adiozdaniel) [reply main-service/README.md:45]: Thanks for this man!
I have synchronized controller mappings in main-service and rate-service with the documented /api/v1 base paths to ensure consistency for API clients.
